test/langtools/tools/javac/patterns/DisambiguateParenthesizedPattern.java

```diff
@@ -0,0 +1,88 @@
+ * @run main/othervm --enable-preview DisambiguateParenthesizedPattern
+ */
+
+import com.sun.source.tree.CaseLabelTree;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.PatternTree;
+import com.sun.source.tree.SwitchTree;
+import com.sun.tools.javac.file.JavacFileManager;
+import com.sun.tools.javac.parser.JavacParser;
+import com.sun.tools.javac.parser.ParserFactory;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.main.Option;
+import com.sun.tools.javac.util.Options;
+import java.nio.charset.Charset;
+
+public class DisambiguateParenthesizedPattern {
+
+    public static void main(String... args) throws Throwable {
+        DisambiguateParenthesizedPattern test = new DisambiguateParenthesizedPattern();
+        test.disambiguationTest("String s",
+                                 ExpressionType.PATTERN);
+        test.disambiguationTest("String s && s.isEmpty()",
+                                 ExpressionType.PATTERN);
+        test.disambiguationTest("(String s)",
+                                 ExpressionType.PATTERN);
+        test.disambiguationTest("((String s))",
+                                 ExpressionType.PATTERN);
+        test.disambiguationTest("(String) s",
+                                 ExpressionType.EXPRESSION);
+        test.disambiguationTest("((String) s)",
+                                 ExpressionType.EXPRESSION);
+        test.disambiguationTest("((0x1))",
+                                 ExpressionType.EXPRESSION);
+    }
+
+    private final ParserFactory factory;
+
+    public DisambiguateParenthesizedPattern() {
+        Context context = new Context();
+        JavacFileManager jfm = new JavacFileManager(context, true, Charset.defaultCharset());
+        Options.instance(context).put(Option.PREVIEW, "");
+        factory = ParserFactory.instance(context);
+    }
+
+    void disambiguationTest(String snippet, ExpressionType expectedType) {
+        String code = """
+                      public class Test {
+                          private void test() {
+                              switch (null) {
+                                  case SNIPPET -> {}
+                              }
+                          }
+                      }
+                      """.replace("SNIPPET", snippet);
+        JavacParser parser = factory.newParser(code, false, false, false);
+        CompilationUnitTree result = parser.parseCompilationUnit();
+        ClassTree clazz = (ClassTree) result.getTypeDecls().get(0);
+        MethodTree method = (MethodTree) clazz.getMembers().get(0);
+        SwitchTree st = (SwitchTree) method.getBody().getStatements().get(0);
+        CaseLabelTree label = st.getCases().get(0).getLabels().get(0);
+        ExpressionType actualType = switch (label) {
```

Comment:
Should the test be careful of using a pattern match switch?

Reply:
I don't think using the new feature in the tests is problematic (esp. javac tests related to the feature). It helps to ensure the feature really works on real code.